src/DraupMembershipERC721.sol

```diff
@@ -0,0 +1,159 @@
+        }
+        _claimed[msg.sender] = true;
+        nextTokenId++;
+        if (nextTokenId > MAX_SUPPLY) {
+            revert MaxSupplyReached();
+        }
+        _mint(msg.sender, nextTokenId - 1);
+    }
+
+    // token trading is disabled initially but will be enabled by the owner
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 firstTokenId,
+        uint256 batchSize
+    ) internal virtual override {
+        if (!transfersAllowed && from != address(0)) {
+            revert TransfersNotAllowed();
+        }
+        super._beforeTokenTransfer(from, to, firstTokenId, batchSize);
+    }
+
+    // on-chain royalty enforcement integration
+    function setApprovalForAll(address operator, bool approved) public override onlyAllowedOperatorApproval(operator) {
+        super.setApprovalForAll(operator, approved);
+    }
+
+    function approve(address operator, uint256 tokenId) public override onlyAllowedOperatorApproval(operator) {
+        super.approve(operator, tokenId);
+    }
+
+    function transferFrom(address from, address to, uint256 tokenId) public override onlyAllowedOperator(from) {
+        super.transferFrom(from, to, tokenId);
+    }
+
+    function safeTransferFrom(address from, address to, uint256 tokenId) public override onlyAllowedOperator(from) {
+        super.safeTransferFrom(from, to, tokenId);
+    }
+
+    function safeTransferFrom(address from, address to, uint256 tokenId, bytes memory data)
+        public
+        override
+        onlyAllowedOperator(from)
+    {
+        super.safeTransferFrom(from, to, tokenId, data);
+    }
+
+    // upgradeable token renderer based on web3-scaffold example by frolic.eth
+    // https://github.com/holic/web3-scaffold/blob/main/packages/contracts/src/IRenderer.sol
+    function tokenURI(uint256 tokenId)
+        public
+        view
+        override
+        returns (string memory)
+    {
+        _requireMinted(tokenId);
+        if (address(renderer) != address(0)) {
+            return renderer.tokenURI(tokenId);
+        }
+        return
+            string(
+                abi.encodePacked(
+                    baseTokenURI,
+                    PaddedString.digitsToString(tokenId, 3),
+                    ".json"
+                )
+            );
+    }
+
+    // Royalty info provided via EIP-2981
+    // https://eips.ethereum.org/EIPS/eip-2981
+    function supportsInterface(bytes4 _interfaceId)
+        public
+        view
+        override
+        returns (bool)
+    {
+        return
+            _interfaceId == type(IERC2981).interfaceId ||
+            super.supportsInterface(_interfaceId);
+    }
+
+    function royaltyInfo(uint256, uint256 salePrice)
+        external
+        view
+        returns (address, uint256)
+    {
+        return (address(this), (salePrice * ROYALTY) / 10000);
+    }
+
+    // Admin actions
+    function setMerkleRoot(bytes32 _merkleRoot) external onlyOwner {
```

Comment:
Definitely a good idea to have an escape hatch in case the merkle root generation breaks off contract! 

Another approach you can take is to have additive Merkle roots by storing a hash map of roots to bools `{merkleRoot: bool}`. You can find the root from the proof and check to see if it's in the hash map. Maybe not worth the complexity, but throwing it out there just in case.

Here's a little helper function that we built at Foundation to do the root lookup.

```
import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";

/**
 * @title Helper library for interacting with Merkle trees & proofs.
 * @author batu-inal & HardlyDifficult & reggieag
 */
library MerkleAddressLibrary {
  using MerkleProof for bytes32[];

  /**
   * @notice Gets the root for a merkle tree comprised only of addresses, using the msg.sender.
   */
  function getMerkleRootForSender(bytes32[] calldata proof) internal view returns (bytes32 root) {
    bytes32 leaf = keccak256(abi.encodePacked(msg.sender));
    root = proof.processProofCalldata(leaf);
  }
}
```